File: agent/python/mango_agent/persistence/persistent_store.py
"""Debounced JSON persistence (FileQueue-style) for the sidecar process."""

from __future__ import annotations

import json
import threading
from pathlib import Path
from typing import Any, Callable, Generic, TypeVar

from mango_agent.persistence.debug import debug

T = TypeVar("T")

DEFAULT_DEBOUNCE_S = 1.0


class PersistentStore(Generic[T]):
    def __init__(
        self,
        *,
        file_path: str | Path,
        serialize: Callable[[T], Any],
        deserialize: Callable[[Any], T | None],
        empty_state: Callable[[], T],
        recover: Callable[[T], T] | None = None,
        debounce_s: float = DEFAULT_DEBOUNCE_S,
        scope: str = "store",
    ) -> None:
        self._file_path = Path(file_path)
        self._serialize = serialize
        self._deserialize = deserialize
        self._empty_state = empty_state
        self._recover = recover
        self._debounce_s = max(0.0, float(debounce_s))
        self._scope = scope
        self._state: T = empty_state()
        self._timer: threading.Timer | None = None
        self._lock = threading.RLock()
        self._destroyed = False

    def get_state(self) -> T:
        with self._lock:
            return self._state

    def set_state(self, next_state: T) -> None:
        with self._lock:
            self._state = next_state
            self.debounce_persist()

    def replace_state(self, next_state: T) -> None:
        self.set_state(next_state)

    def load_from_storage(self) -> T:
        with self._lock:
            try:
                if not self._file_path.is_file():
                    self._state = self._empty_state()
                    return self._state
                raw_text = self._file_path.read_text(encoding="utf-8")
                if not raw_text.strip():
                    self._state = self._empty_state()
                    return self._state
                parsed = json.loads(raw_text)
                loaded = self._deserialize(parsed)
                next_state = loaded if loaded is not None else self._empty_state()
                if self._recover is not None:
                    next_state = self._recover(next_state)
                self._state = next_state
                return self._state
            except Exception as exc:  # noqa: BLE001
                debug(self._scope, "load_from_storage failed", exc)
                self._state = self._empty_state()
                return self._state

    def persist_now(self) -> None:
        with self._lock:
            if self._timer is not None:
                self._timer.cancel()
                self._timer = None
            try:
                self._file_path.parent.mkdir(parents=True, exist_ok=True)
                payload = json.dumps(self._serialize(self._state), ensure_ascii=False, indent=2)
                tmp = self._file_path.with_suffix(self._file_path.suffix + ".tmp")
                tmp.write_text(payload, encoding="utf-8")
                tmp.replace(self._file_path)
            except Exception as exc:  # noqa: BLE001
                debug(self._scope, "persist_now failed", exc)

    def debounce_persist(self) -> None:
        with self._lock:
            if self._destroyed:
                return
            if self._debounce_s <= 0:
                self.persist_now()
                return
            if self._timer is not None:
                self._timer.cancel()
            timer = threading.Timer(self._debounce_s, self._persist_from_timer)
            timer.daemon = True
            self._timer = timer
            timer.start()

    def destroy(self) -> None:
        with self._lock:
            self._destroyed = True
            self.persist_now()

    def _persist_from_timer(self) -> None:
        with self._lock:
            self._timer = None
            if self._destroyed:
                return
            self.persist_now()

File: apps/electron/src/main/lib/persistent-store.ts
/**
 * Debounced JSON persistence (FileQueue-style).
 * In-memory state + loadFromStorage / persistNow / debouncePersist / destroy.
 */
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";
import { debug } from "./debug";

const DEFAULT_DEBOUNCE_MS = 1000;

export type PersistentStoreOptions<T> = {
  filePath: string;
  debounceMs?: number;
  /** Serialize in-memory state for disk. */
  serialize: (state: T) => unknown;
  /** Deserialize disk payload; return null to use emptyState. */
  deserialize: (raw: unknown) => T | null;
  emptyState: () => T;
  /** Optional crash-recovery / normalize after load. */
  recover?: (state: T) => T;
  scope?: string;
};

export class PersistentStore<T> {
  private state: T;
  private persistDebounceTimer: ReturnType<typeof setTimeout> | null = null;
  private readonly filePath: string;
  private readonly debounceMs: number;
  private readonly serialize: (state: T) => unknown;
  private readonly deserialize: (raw: unknown) => T | null;
  private readonly emptyState: () => T;
  private readonly recover?: (state: T) => T;
  private readonly scope: string;
  private destroyed = false;

  constructor(options: PersistentStoreOptions<T>) {
    this.filePath = options.filePath;
    this.debounceMs = options.debounceMs ?? DEFAULT_DEBOUNCE_MS;
    this.serialize = options.serialize;
    this.deserialize = options.deserialize;
    this.emptyState = options.emptyState;
    this.recover = options.recover;
    this.scope = options.scope ?? "store";
    this.state = this.emptyState();
  }

  getState(): T {
    return this.state;
  }

  setState(next: T): void {
    this.state = next;
    this.debouncePersist();
  }

  /** Replace state and schedule debounced write. */
  replaceState(next: T): void {
    this.setState(next);
  }

  loadFromStorage(): T {
    try {
      if (!existsSync(this.filePath)) {
        this.state = this.emptyState();
        return this.state;
      }
      const rawText = readFileSync(this.filePath, "utf8");
      if (!rawText.trim()) {
        this.state = this.emptyState();
        return this.state;
      }
      const parsed = JSON.parse(rawText) as unknown;
      const loaded = this.deserialize(parsed);
      let next = loaded ?? this.emptyState();
      if (this.recover) {
        next = this.recover(next);
      }
      this.state = next;
      return this.state;
    } catch (error) {
      debug(this.scope, "loadFromStorage failed", error);
      this.state = this.emptyState();
      return this.state;
    }
  }

  persistNow(): void {
    if (this.persistDebounceTimer) {
      clearTimeout(this.persistDebounceTimer);
      this.persistDebounceTimer = null;
    }
    try {
      mkdirSync(dirname(this.filePath), { recursive: true });
      const payload = JSON.stringify(this.serialize(this.state), null, 2);
      writeFileSync(this.filePath, payload, "utf8");
    } catch (error) {
      debug(this.scope, "persistNow failed", error);
    }
  }

  debouncePersist(): void {
    if (this.destroyed) return;
    if (this.debounceMs <= 0) {
      this.persistNow();
      return;
    }
    if (this.persistDebounceTimer) {
      clearTimeout(this.persistDebounceTimer);
    }
    this.persistDebounceTimer = setTimeout(() => {
      this.persistDebounceTimer = null;
      this.persistNow();
    }, this.debounceMs);
  }

  /** Flush pending timer and write synchronously. */
  destroy(): void {
    this.destroyed = true;
    this.persistNow();
  }
}
